state/entrysync.go

```diff
@@ -0,0 +1,313 @@
+			if es.has(ask.hash) {
+				es.asks = append(es.asks[:i], es.asks[i+1:]...)
+				i-- // we removed one, index stays the same
+				delete(es.askMap, ask.hash.Fixed())
+				continue
+			}
+
+			// re-ask
+			if time.Since(ask.time) > EntrySyncRetry {
+				request := messages.NewMissingData(primitives.NewTimestampNow(), ask.hash)
+				request.SendOut(es.s, request)
+				ask.time = time.Now()
+				ask.tries++
+
+				if ask.tries%EntrySyncWarning == 0 {
+					es.s.Logger.WithField("tries", ask.tries).WithField("hash", ask.hash.String()).Warnf("Unable to retrieve entry from network")
+				}
+			}
+		}
+		es.askMtx.Unlock()
+
+		time.Sleep(time.Second)
+	}
+}
+
+// returns the maximum possible height we have eblocks for
+func (es *EntrySync) syncMax() uint32 {
+	end := es.s.GetHighestSavedBlk()
+	if es.s.DBStates.ProcessHeight < end {
+		end = es.s.DBStates.ProcessHeight
+	}
+	return end
+}
+
+// SyncHeight starts the Sync, Ask, and Check routines.
+func (es *EntrySync) SyncHeight() {
+	go es.ask()
+	go es.check()
+
+	es.position = es.s.EntryDBHeightComplete
+	// genesis block edge case since EntryDBHeightComplete can't be -1
+	if es.position > 0 {
+		es.position++
+	}
+
+	for {
+		select {
+		case <-es.closer:
+			return
+		default:
+		}
+		// block not available yet
+		if es.position > es.syncMax() {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		db := es.s.GetDirectoryBlockByHeight(es.position)
+		if db == nil { // block not saved yet
+			time.Sleep(time.Second)
+			continue
+		}
+
+		// legacy logic
+		if !es.s.DBFinished { // throttle syncing
+			time.Sleep(time.Millisecond * 10)
+		}
+
+		es.syncDBlock(db)
+		es.position++
+	}
+}
+
+// extract eblocks from the dblock
+func (es *EntrySync) syncDBlock(db interfaces.IDirectoryBlock) {
+	eblocks := db.GetEntryHashes()[3:] // skip f/c/a-block
+	if len(eblocks) > 0 {
+		for _, keymr := range eblocks {
+			for !es.syncEBlock(db.GetDatabaseHeight(), keymr, db.GetTimestamp()) {
+				time.Sleep(time.Second)
+			}
+		}
+	} else {
+		es.syncNoEBlock(db.GetDatabaseHeight())
+	}
+}
+
+// for dblocks with no eblocks, we still need to advance the complete counter
+func (es *EntrySync) syncNoEBlock(height uint32) {
+	ebsync := new(entrySyncEBlock)
+	ebsync.height = height
+	es.ebMtx.Lock()
+	es.eblocks = append(es.eblocks, ebsync)
+	es.ebMtx.Unlock()
+}
+
+// process a single eblock, will call syncEntryHash() on each entry, and add an eblock to the queue
+func (es *EntrySync) syncEBlock(height uint32, keymr interfaces.IHash, ts interfaces.Timestamp) bool {
+	eblock, err := es.s.DB.FetchEBlock(keymr)
+	if err != nil { // database corrupt
+		panic(err)
+	}
+
+	if eblock == nil {
+		return false
+	}
+
+	ebsync := new(entrySyncEBlock)
+	ebsync.height = height
+
+	for _, entryHash := range eblock.GetEntryHashes() {
+		if entryHash.IsMinuteMarker() {
+			continue
+		}
+
+		// see state.UpdateState()
+		update := new(EntryUpdate)
+		update.Hash = entryHash
+		update.Timestamp = ts
+		es.s.UpdateEntryHash <- update
```

Comment:
NOTE: need to check that

Reply:
It's the following code:
https://github.com/FactomProject/factomd/blob/ba6800ab3ba2296acf82726af2bd64d65c88b7ca/state/state.go#L1375-L1388

Reply:
ok thanks!